{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48c7c36e-668d-43ad-9ca0-97bfba3275f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "sys.path.append('..')\n",
    "from tools.formating.formating import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa098528-6554-4d4e-872a-1b5493f301d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the absolute path of the desired file\n",
    "r_path = os.path.abspath(\"../tools/formating/formating.R\")\n",
    "\n",
    "with open(r_path, 'r') as r_source_file:\n",
    "    r_source = r_source_file.read()\n",
    "\n",
    "# Evaluate the R script in the R environment\n",
    "ro.r(r_source)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f55472f-2449-4420-81b3-9ffc4c532a82",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f418058-cd75-4108-9049-179fb28c2dde",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"/ps/ai-ready/data/Clustering/Tabula_Muris/droplet_Bladder_seurat_tiss.Robj\"\n",
    "LoadSeurat_r = ro.globalenv['LoadSeurat']\n",
    "srat = LoadSeurat_r(path)\n",
    "srat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc0a1276-630d-481b-a4c1-036bdc7e6a7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the absolute path of the desired file\n",
    "r_path = os.path.abspath(\"../tools/qc/seurat_qc.R\")\n",
    "\n",
    "with open(r_path, 'r') as r_source_file:\n",
    "    r_source = r_source_file.read()\n",
    "\n",
    "# Evaluate the R script in the R environment\n",
    "ro.r(r_source)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce7c63b5-da2d-4511-944a-89bf36ca1383",
   "metadata": {},
   "outputs": [],
   "source": [
    "LoadSeuratMetaData_r = ro.globalenv['LoadSeuratMetaData']\n",
    "results = LoadSeuratMetaDataC_r(path, '/ps/ai-ready/data/Clustering/Tabula_Muris/droplet_Bladder_seurat_tiss.h5Seurat')\n",
    "srat <- list(results)[0]\n",
    "srat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2c0c783-bf62-4821-a655-61b7db88ce67",
   "metadata": {},
   "outputs": [],
   "source": [
    "RunSeuratQC_r = ro.globalenv['RunSeuratQC']\n",
    "results = RunSeuratQC_r(srat, '/ps/ai-ready/data/Clustering/Tabula_Muris/droplet_Bladder_seurat_tiss.h5Seurat')\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba22a3b9-caf9-4241-8d43-669e488bb2e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"/ps/ai-ready/data/Clustering/Tabula_Muris/droplet_Bladder_seurat_tiss.Robj\"\n",
    "# default_assay, assay_names, metadata, nCells, nGenes, genes, cells, HVGsID, pca, tsne, umap = GetMetadataFromSeurat(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e919f06e-c8a4-438c-9c82-0217891a9c7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"/ps/ai-ready/data/Clustering/Tabula_Muris/droplet_Bladder_seurat_tiss.Robj\"\n",
    "RunSeuratQC_r = ro.globalenv['RunSeuratQC']\n",
    "results = RunSeuratQC_r(path, '/ps/ai-ready/data/Clustering/Tabula_Muris/droplet_Bladder_seurat_tiss.h5Seurat')\n",
    "results\n",
    "# default_assay = list(results[0])[0]\n",
    "# assay_names = list(results[1])\n",
    "# nCells = list(results[3])[0]\n",
    "# nGenes = list(results[4])[0]\n",
    "# genes  = list(results[5])\n",
    "# cells = list(results[6])\n",
    "# if results[7] != ro.rinterface.NULL:\n",
    "#     HVGsID = list(results[7])\n",
    "    \n",
    "#     with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "#         if results[2] != ro.rinterface.NULL: \n",
    "#             metadata = ro.conversion.rpy2py(results[2])\n",
    "#         if results[8] != ro.rinterface.NULL:\n",
    "#             pca = ro.conversion.rpy2py(results[8])\n",
    "#         if results[9] != ro.rinterface.NULL:\n",
    "#             tsne = ro.conversion.rpy2py(results[9])\n",
    "#         if results[10] != ro.rinterface.NULL:\n",
    "#             umap = ro.conversion.rpy2py(results[10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89690228-6663-4466-bb33-5866429d1cd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41711801-eeb8-40e2-9a83-89363b9cbe07",
   "metadata": {},
   "outputs": [],
   "source": [
    "default_assay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31e6380b-8fb5-4645-9b4d-35c4f63a062c",
   "metadata": {},
   "outputs": [],
   "source": [
    "assay_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02ebaa59-f059-4ad7-b37f-def4165d0a5e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "nCells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acc0f0ac-6819-4105-b658-48e3c9e3aa22",
   "metadata": {},
   "outputs": [],
   "source": [
    "nGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e9eac1a-3547-4938-bc22-5e378d9bdfc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c038af06-1f4d-4e45-85c1-deb1dcec8142",
   "metadata": {},
   "outputs": [],
   "source": [
    "cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb1652ea-7c03-480a-8756-1179307e41c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "HVGsID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ba33f19-1339-4a0c-9dda-d9c9c43f670c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "571d6c19-9a0f-4b4b-b0be-0fdcb4e34510",
   "metadata": {},
   "outputs": [],
   "source": [
    "tsne"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eb97f40-45a4-43b6-8a8a-22c4ec8375bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "type(umap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cfcb73e-e8bb-42cb-8e89-1502ac304ba5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from rpy2 import robjects as robj\n",
    "umap == robj.rinterface.NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a1b300d-9c50-4a12-871a-b449acb5ba41",
   "metadata": {},
   "outputs": [],
   "source": [
    "GetMetadataFromSeurat_r = ro.globalenv['GetMetadataFromSeurat']\n",
    "results = GetMetadataFromSeurat_r(path)\n",
    "results = list(results)\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6957289a-eadf-446e-b0cd-1afdbbe72cce",
   "metadata": {},
   "outputs": [],
   "source": [
    "results[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67efd901-40c6-4896-ae93-045cb9369f6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "    metadata = ro.conversion.rpy2py(results[2])\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6b3cd80-197e-4371-ba8d-b2e9659891b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "results[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4a4d629-ad4c-493a-925c-2834504f04ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = ro.r.seq(1, 10)\n",
    "print(x.ro)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0801c23d-5676-4528-8dd1-178e16a34665",
   "metadata": {},
   "outputs": [],
   "source": [
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c04e0682-3f86-49ae-829a-2a45a853fc06",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: SummarizedExperiment\n",
      "\n",
      "R[write to console]: Loading required package: MatrixGenerics\n",
      "\n",
      "R[write to console]: Loading required package: matrixStats\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘celldex’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:SingleR’:\n",
      "\n",
      "    BlueprintEncodeData, DatabaseImmuneCellExpressionData,\n",
      "    HumanPrimaryCellAtlasData, ImmGenData, MonacoImmuneData,\n",
      "    MouseRNAseqData, NovershternHematopoieticData\n",
      "\n",
      "\n",
      "R[write to console]: here() starts at /ps/Machine-learning-development-environment-for-single-cell-sequencing-data-analyses\n",
      "\n",
      "R[write to console]: Loading required package: SingleCellExperiment\n",
      "\n",
      "R[write to console]: Loading required package: scuttle\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,\n",
      "which was just loaded, will retire in October 2023.\n",
      "Please refer to R-spatial evolution reports for details, especially\n",
      "https://r-spatial.org/r/2023/05/15/evolution4.html.\n",
      "It may be desirable to make the sf package available;\n",
      "package maintainers should consider adding sf to Suggests:.\n",
      "The sp package is now running under evolution status 2\n",
      "     (status 2 uses the sf package in place of rgdal)\n",
      "\n",
      "R[write to console]: Attaching SeuratObject\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "R[write to console]: Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "sys.path.append('..')\n",
    "from tools.qc.seurat_qc import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9d2e088a-682e-4ecf-a1fc-9eebc2f0d966",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Updating from v2.X to v3.X\n",
      "\n",
      "R[write to console]: Validating object structure\n",
      "\n",
      "R[write to console]: Updating object slots\n",
      "\n",
      "R[write to console]: Ensuring keys are in the proper structure\n",
      "\n",
      "R[write to console]: Ensuring feature names don't have underscores or pipes\n",
      "\n",
      "R[write to console]: Object representation is consistent with the most current Seurat version\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"RNA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Performing log-normalization\n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n",
      "R[write to console]: Calculating gene variances\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n",
      "R[write to console]: Calculating feature variances of standardized and clipped values\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n",
      "R[write to console]: Centering and scaling data matrix\n",
      "\n",
      "  |                                                                            \n",
      "  |                                                                      |   0%\n",
      "  |                                                                            \n",
      "  |===                                                                   |   4%\n",
      "  |                                                                            \n",
      "  |======                                                                |   8%\n",
      "  |                                                                            \n",
      "  |=========                                                             |  12%\n",
      "  |                                                                            \n",
      "  |============                                                          |  17%\n",
      "  |                                                                            \n",
      "  |===============                                                       |  21%\n",
      "  |                                                                            \n",
      "  |==================                                                    |  25%\n",
      "  |                                                                            \n",
      "  |====================                                                  |  29%\n",
      "  |                                                                            \n",
      "  |=======================                                               |  33%\n",
      "  |                                                                            \n",
      "  |==========================                                            |  38%\n",
      "  |                                                                            \n",
      "  |=============================                                         |  42%\n",
      "  |                                                                            \n",
      "  |================================                                      |  46%\n",
      "  |                                                                            \n",
      "  |===================================                                   |  50%\n",
      "  |                                                                            \n",
      "  |======================================                                |  54%\n",
      "  |                                                                            \n",
      "  |=========================================                             |  58%\n",
      "  |                                                                            \n",
      "  |============================================                          |  62%\n",
      "  |                                                                            \n",
      "  |===============================================                       |  67%\n",
      "  |                                                                            \n",
      "  |==================================================                    |  71%\n",
      "  |                                                                            \n",
      "  |====================================================                  |  75%\n",
      "  |                                                                            \n",
      "  |=======================================================               |  79%\n",
      "  |                                                                            \n",
      "  |==========================================================            |  83%\n",
      "  |                                                                            \n",
      "  |=============================================================         |  88%\n",
      "  |                                                                            \n",
      "  |================================================================      |  92%\n",
      "  |                                                                            \n",
      "  |===================================================================   |  96%\n",
      "  |                                                                            \n",
      "  |======================================================================| 100%\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: PC_ 1 \n",
      "Positive:  Foxq1, Mal, Ly6d, Wfdc2, Fxyd3, Krt18, Aqp3, Gsta4, 2200002D01Rik, Spint2 \n",
      "\t   Krt7, Sfn, Perp, Krt19, Krt8, 1110032A04Rik, Gsto1, Cldn4, Sprr1a, Ezr \n",
      "\t   Upk1a, Gstm1, Igfbp2, Lmo1, Sdc1, Cldn7, Cdh1, Ivl, Mgst3, Ctse \n",
      "Negative:  Lgals1, Vim, Serpinh1, Sparc, Serping1, Col1a2, Bgn, Col3a1, Col1a1, Dcn \n",
      "\t   Plac9, Mgp, Rarres2, Col6a1, Htra1, Rcn3, Nbl1, Tuba1a, Plac8, Pcolce \n",
      "\t   Col6a2, Id3, Gadd45g, Fstl1, Mxra8, Cd302, Lsp1, Cebpd, Ppic, Cst3 \n",
      "PC_ 2 \n",
      "Positive:  Tyrobp, Ctss, H2-Eb1, H2-Aa, Cd52, Laptm5, Fcer1g, Cd74, Ly86, Cd83 \n",
      "\t   H2-DMa, Sfpi1, Cd53, H2-Ab1, Pld4, Plek, Lyz2, Coro1a, Plbd1, Srgn \n",
      "\t   Lyz1, Alox5ap, H2-DMb1, Bcl2a1b, Bcl2a1d, Rac2, Ms4a6c, Il1b, Clec12a, Gmfg \n",
      "Negative:  S100a6, Serpinh1, Col3a1, Dcn, Col6a1, Col1a1, Ppap2b, Htra1, Mgp, Col1a2 \n",
      "\t   Serping1, Rcn3, Pcolce, Dkk3, Gsn, Ecm1, Mxra8, Bgn, Fbln5, Sparc \n",
      "\t   Rarres2, Nbl1, Col6a2, Col5a2, Ppic, Mmp2, Fstl1, Aebp1, Prelp, Col6a3 \n",
      "PC_ 3 \n",
      "Positive:  Tgfbi, Car3, Rbp4, Tnc, Basp1, Spon1, Col6a3, Gpx3, Gm2a, Hhip \n",
      "\t   Thbs2, Col12a1, Cxcl14, Mmp2, Cpz, A130040M12Rik, 3632451O06Rik, Aldh1a2, Hk2, Fn1 \n",
      "\t   H2-DMb1, Itga11, Kcnk2, Fam105a, Col11a1, Creb3l1, Dkk2, Bmp2, Cpxm1, H2-Ab1 \n",
      "Negative:  Pecam1, Esam, Egfl7, Cdh5, Emcn, Podxl, Eltd1, Pcp4l1, S1pr1, Cd93 \n",
      "\t   Gimap6, Mmrn2, Flt1, C130074G19Rik, Gpr116, Jam2, Sox17, Rasip1, Robo4, Kdr \n",
      "\t   Ecscr, Abcb1a, Plvap, Ldb2, Ushbp1, Apold1, Grrp1, Myct1, Epas1, Scarb1 \n",
      "PC_ 4 \n",
      "Positive:  Car3, Tnc, Rbp4, Hhip, Tspan13, Cxcl14, Cpz, Myl9, Dkk2, 3632451O06Rik \n",
      "\t   Pde4b, Gpx3, Itga11, Aldh1a2, Spon1, Gas6, Tgfbi, Col11a1, Col4a2, Plk2 \n",
      "\t   Kcnk2, Tagln, Acta2, Col6a3, Wfdc1, Cd200, Thy1, Notum, Procr, Palld \n",
      "Negative:  Scara5, Clec3b, Mfap5, Pi16, Krtdap, Cd34, Gsn, Sparcl1, Abi3bp, Col14a1 \n",
      "\t   Npy1r, Tnxb, Lum, Lbp, Ccdc80, Timp3, Sfrp1, Inmt, Fbln1, Atp1a2 \n",
      "\t   Dmkn, Chodl, Htra3, Col8a1, Sbsn, Osr1, Gfpt2, Fxyd5, Gpr133, Igfbp4 \n",
      "PC_ 5 \n",
      "Positive:  Bcam, Gsdmc2, Has3, Slc7a11, Col17a1, Itga3, Kcnn4, Nrn1, Srxn1, 2310043J07Rik \n",
      "\t   Trf, Aldh3a1, Ank, Cebpb, Gsdmc3, Vegfa, A130040M12Rik, Tns4, Serpinb5, Dusp6 \n",
      "\t   Tinagl1, Trp63, Wdr92, Krt15, Phospho1, Dppa5a, Gclc, Gsr, Gsta3, Rgs10 \n",
      "Negative:  Psca, Fam25c, Pdzk1ip1, Upk2, Acer2, Snhg11, Upk3a, Galntl1, AA986860, Sprr2b \n",
      "\t   Cyp4f15, Defb1, Snx31, Tspan13, Mmp15, Fmo5, Sgpl1, Cldn23, Pttg1ip, Tmem64 \n",
      "\t   Gjb6, Myh14, Hebp2, Fut9, Tmem123, Vamp5, Wfdc15b, Capn5, Arg2, Ppfibp2 \n",
      "\n",
      "R[write to console]: Computing nearest neighbor graph\n",
      "\n",
      "R[write to console]: Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 2500\n",
      "Number of edges: 73418\n",
      "\n",
      "Running Louvain algorithm...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Maximum modularity in 10 random starts: 0.8868\n",
      "Number of communities: 10\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Warning:\n",
      "R[write to console]:  The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
      "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
      "This message will be shown once per session\n",
      "\n",
      "R[write to console]: 04:00:02 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "R[write to console]: 04:00:02 Read 2500 rows and found 10 numeric columns\n",
      "\n",
      "R[write to console]: 04:00:02 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "R[write to console]: 04:00:02 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n",
      "R[write to console]: 04:00:03 Writing NN index file to temp file /tmp/RtmpY044Ul/file1530fcf69c\n",
      "\n",
      "R[write to console]: 04:00:03 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "R[write to console]: 04:00:03 Annoy recall = 100%\n",
      "\n",
      "R[write to console]: 04:00:04 Commencing smooth kNN distance calibration using 1 thread\n",
      "R[write to console]:  with target n_neighbors = 30\n",
      "\n",
      "R[write to console]: 04:00:06 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "R[write to console]: 04:00:07 Commencing optimization for 500 epochs, with 93296 positive edges\n",
      "\n",
      "R[write to console]: Using method '\n",
      "R[write to console]: umap\n",
      "R[write to console]: '\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n",
      "R[write to console]: 04:00:10 Optimization finished\n",
      "\n",
      "R[write to console]: Warning:\n",
      "R[write to console]:  Overwriting previous file /ps/ai-ready/data/Clustering/Tabula_Muris/droplet_Bladder_seurat_tiss.h5Seurat\n",
      "\n",
      "R[write to console]: Creating h5Seurat file for version 3.1.5.9900\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Seurat object is saved successfully.\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Validating h5Seurat file\n",
      "\n",
      "R[write to console]: Adding scale.data from RNA as X\n",
      "\n",
      "R[write to console]: Transfering meta.features to var\n",
      "\n",
      "R[write to console]: Adding data from RNA as raw\n",
      "\n",
      "R[write to console]: Transfering meta.features to raw/var\n",
      "\n",
      "R[write to console]: Transfering meta.data to obs\n",
      "\n",
      "R[write to console]: Adding dimensional reduction information for pca\n",
      "\n",
      "R[write to console]: Adding feature loadings for pca\n",
      "\n",
      "R[write to console]: Adding dimensional reduction information for tsne\n",
      "\n",
      "R[write to console]: Adding dimensional reduction information for umap\n",
      "\n",
      "R[write to console]: Adding RNA_snn as neighbors\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"AnnData object is saved successfully.\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'RNA'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = \"/ps/ai-ready/data/Clustering/Tabula_Muris/droplet_Bladder_seurat_tiss.Robj\"\n",
    "output_path = '/ps/ai-ready/data/Clustering/Tabula_Muris/droplet_Bladder_seurat_tiss.h5Seurat'\n",
    "default_assay, assay_names, metadata, nCells, nGenes, genes, cells, HVGsID, pca, tsne, umap, adata_path=run_seurat_qc(path, output_path)\n",
    "default_assay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8962281f-bf44-4d97-bd9f-7449a5ee153c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/ps/ai-ready/data/Clustering/Tabula_Muris/droplet_Bladder_seurat_tiss.h5ad'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2736ee37-4bf4-45a1-87d3-ecf758849f29",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
