{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b6120aea-a550-40ec-93f0-64888aa90448",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import rpy2.robjects as ro\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects import pandas2ri\n",
    "from rpy2.robjects.conversion import localconverter\n",
    "# from tools.formating.formating import change_file_extension, load_anndata, convert_from_r, convert_to_r\n",
    "pandas2ri.activate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c27056a4-7ff1-4178-98bf-7422146e022e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: SummarizedExperiment\n",
      "\n",
      "R[write to console]: Loading required package: MatrixGenerics\n",
      "\n",
      "R[write to console]: Loading required package: matrixStats\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘celldex’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:SingleR’:\n",
      "\n",
      "    BlueprintEncodeData, DatabaseImmuneCellExpressionData,\n",
      "    HumanPrimaryCellAtlasData, ImmGenData, MonacoImmuneData,\n",
      "    MouseRNAseqData, NovershternHematopoieticData\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: SingleCellExperiment\n",
      "\n",
      "R[write to console]: Loading required package: scuttle\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Attaching SeuratObject\n",
      "\n",
      "R[write to console]: Seurat v4 was just loaded with SeuratObject v5; disabling v5 assays and\n",
      "validation routines, and ensuring assays work in strict v3/v4\n",
      "compatibility mode\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘Seurat’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "R[write to console]: Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "\n",
      "R[write to console]: Signac built for for SeuratObject v4 was just loaded with SeuratObject\n",
      "v5; disabling v5 assays and validation routines, and ensuring assays\n",
      "work in strict v3/v4 compatibility mode\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘Matrix’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "R[write to console]: X -> counts\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'Seurat' [package \"SeuratObject\"] with 13 slots\n",
      "  ..@ assays      :List of 1\n",
      "  .. ..$ RNA:Formal class 'Assay' [package \"SeuratObject\"] with 8 slots\n",
      "  .. .. .. ..@ counts       :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int [1:2957496] 26 39 52 53 54 78 122 152 168 178 ...\n",
      "  .. .. .. .. .. ..@ p       : int [1:3437] 0 3033 3600 4041 4764 5192 5642 6295 6838 8098 ...\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 33538 3436\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:33538] \"MIR1302-2HG\" \"FAM138A\" \"OR4F5\" \"AL627309.1\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr [1:3436] \"AACCCAAAGTACGAGC-1-H0015_apex\" \"AACGAAACACGACTAT-1-H0015_apex\" \"AAGTGAATCGAGAGCA-1-H0015_apex\" \"AAGTGAATCTGATTCT-1-H0015_apex\" ...\n",
      "  .. .. .. .. .. ..@ x       : num [1:2957496] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ data         :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int [1:2957496] 26 39 52 53 54 78 122 152 168 178 ...\n",
      "  .. .. .. .. .. ..@ p       : int [1:3437] 0 3033 3600 4041 4764 5192 5642 6295 6838 8098 ...\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 33538 3436\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:33538] \"MIR1302-2HG\" \"FAM138A\" \"OR4F5\" \"AL627309.1\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr [1:3436] \"AACCCAAAGTACGAGC-1-H0015_apex\" \"AACGAAACACGACTAT-1-H0015_apex\" \"AAGTGAATCGAGAGCA-1-H0015_apex\" \"AAGTGAATCTGATTCT-1-H0015_apex\" ...\n",
      "  .. .. .. .. .. ..@ x       : num [1:2957496] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ scale.data   : num[0 , 0 ] \n",
      "  .. .. .. ..@ assay.orig   : NULL\n",
      "  .. .. .. ..@ var.features : logi(0) \n",
      "  .. .. .. ..@ meta.features:'data.frame':\t33538 obs. of  0 variables\n",
      "  .. .. .. ..@ misc         : list()\n",
      "  .. .. .. ..@ key          : chr \"rna_\"\n",
      "  ..@ meta.data   :'data.frame':\t3436 obs. of  29 variables:\n",
      "  .. ..$ orig.ident     : Factor w/ 1 level \"hca_heart_neuronal_raw\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ nCount_RNA     : num [1:3436] 5796 743 562 984 583 ...\n",
      "  .. ..$ nFeature_RNA   : int [1:3436] 3033 567 441 723 428 450 653 543 1260 1277 ...\n",
      "  .. ..$ nCounts_RNA    : num [1:3436] 5796 743 562 984 583 ...\n",
      "  .. ..$ nFeatures_RNA  : int [1:3436] 3033 567 441 723 428 450 653 543 1260 1277 ...\n",
      "  .. ..$ NRP            : Factor w/ 2 levels \"No\",\"Yes\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ age_group      : Factor w/ 7 levels \"40-45\",\"45-50\",..: 3 3 3 3 3 3 3 3 3 3 ...\n",
      "  .. ..$ cell_source    : Factor w/ 4 levels \"Harvard-Nuclei\",..: 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ cell_type      : Factor w/ 1 level \"Neuronal\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ donor          : Factor w/ 14 levels \"D1\",\"D2\",\"D3\",..: 12 12 12 12 12 12 12 12 12 12 ...\n",
      "  .. ..$ gender         : Factor w/ 2 levels \"Female\",\"Male\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ percent_mito   : num [1:3436] 0.00414 0.00808 0.00534 0.00407 0.00172 ...\n",
      "  .. ..$ percent_ribo   : num [1:3436] 0.0019 0.00135 0.00534 0 0.00515 ...\n",
      "  .. ..$ region         : Factor w/ 6 levels \"AX\",\"LA\",\"LV\",..: 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ sample         : Factor w/ 133 levels \"H0015_LA_new\",..: 5 5 5 5 5 5 5 5 5 5 ...\n",
      "  .. ..$ scrublet_score : num [1:3436] 0.1393 0.0152 0.0143 0.0152 0.0577 ...\n",
      "  .. ..$ source         : Factor w/ 3 levels \"CD45+\",\"Cells\",..: 3 3 3 3 3 3 3 3 3 3 ...\n",
      "  .. ..$ type           : Factor w/ 2 levels \"DBD\",\"DCD\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ version        : Factor w/ 2 levels \"V2\",\"V3\": 2 2 2 2 2 2 2 2 2 2 ...\n",
      "  .. ..$ cell_states    : Factor w/ 6 levels \"NC1\",\"NC2\",\"NC3\",..: 3 1 1 1 1 1 1 1 1 2 ...\n",
      "  .. ..$ Used           : Factor w/ 1 level \"Yes\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..$ percent.mt     : num [1:3436] 0.414 0.808 0.534 0.407 0.172 ...\n",
      "  .. ..$ percent.rb     : num [1:3436] 0.276 0.404 0.89 0.102 0.858 ...\n",
      "  .. ..$ percent.hb     : num [1:3436] 0 0.135 0 0 0 ...\n",
      "  .. ..$ percent.plat   : num [1:3436] 0.0518 0 0 0.2033 0.1715 ...\n",
      "  .. ..$ RNA_snn_res.0.5: Factor w/ 10 levels \"0\",\"1\",\"2\",\"3\",..: 4 1 1 1 1 1 2 1 1 8 ...\n",
      "  .. ..$ seurat_clusters: Factor w/ 10 levels \"0\",\"1\",\"2\",\"3\",..: 4 1 1 1 1 1 2 1 1 8 ...\n",
      "  .. ..$ doublet_score  : num [1:3436] 0.39462 0.00673 0 0.06726 0 ...\n",
      "  .. ..$ doublet_class  : Factor w/ 1 level \"Singlet\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  ..@ active.assay: chr \"RNA\"\n",
      "  ..@ active.ident: Factor w/ 1 level \"hca_heart_neuronal_raw\": 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. ..- attr(*, \"names\")= chr [1:3436] \"AACCCAAAGTACGAGC-1-H0015_apex\" \"AACGAAACACGACTAT-1-H0015_apex\" \"AAGTGAATCGAGAGCA-1-H0015_apex\" \"AAGTGAATCTGATTCT-1-H0015_apex\" ...\n",
      "  ..@ graphs      : list()\n",
      "  ..@ neighbors   : list()\n",
      "  ..@ reductions  : list()\n",
      "  ..@ images      : list()\n",
      "  ..@ project.name: chr \"hca_heart_neuronal_raw\"\n",
      "  ..@ misc        : list()\n",
      "  ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1\n",
      "  .. ..$ : int [1:3] 5 0 1\n",
      "  ..@ commands    : list()\n",
      "  ..@ tools       : list()\n",
      "NULL\n",
      "[1] \"RNA\"\n",
      "[1] \"Setting default_assay to RNA\"\n",
      "[1] \"Calculating the percentage of mitocondrial per cell and add to the metadata.\"\n",
      "[1] \"Calculating the proportion gene expression that comes from ribosomal proteins.\"\n",
      "[1] \"Calculating the percentage hemoglobin genes - includes all genes starting with HB except HBP.\"\n",
      "[1] \"Filtering low quality genes/cells.\"\n",
      "[1] \"Normalizing dataset using logCP10k.\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Performing log-normalization\n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Finding variable features.\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Calculating gene variances\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n",
      "R[write to console]: Calculating feature variances of standardized and clipped values\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Scaling dataset.\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Centering and scaling data matrix\n",
      "\n",
      "  |                                                                            \n",
      "  |                                                                      |   0%\n",
      "  |                                                                            \n",
      "  |==                                                                    |   3%\n",
      "  |                                                                            \n",
      "  |====                                                                  |   6%\n",
      "  |                                                                            \n",
      "  |======                                                                |   9%\n",
      "  |                                                                            \n",
      "  |========                                                              |  12%\n",
      "  |                                                                            \n",
      "  |==========                                                            |  15%\n",
      "  |                                                                            \n",
      "  |============                                                          |  18%\n",
      "  |                                                                            \n",
      "  |==============                                                        |  21%\n",
      "  |                                                                            \n",
      "  |================                                                      |  24%\n",
      "  |                                                                            \n",
      "  |===================                                                   |  26%\n",
      "  |                                                                            \n",
      "  |=====================                                                 |  29%\n",
      "  |                                                                            \n",
      "  |=======================                                               |  32%\n",
      "  |                                                                            \n",
      "  |=========================                                             |  35%\n",
      "  |                                                                            \n",
      "  |===========================                                           |  38%\n",
      "  |                                                                            \n",
      "  |=============================                                         |  41%\n",
      "  |                                                                            \n",
      "  |===============================                                       |  44%\n",
      "  |                                                                            \n",
      "  |=================================                                     |  47%\n",
      "  |                                                                            \n",
      "  |===================================                                   |  50%\n",
      "  |                                                                            \n",
      "  |=====================================                                 |  53%\n",
      "  |                                                                            \n",
      "  |=======================================                               |  56%\n",
      "  |                                                                            \n",
      "  |=========================================                             |  59%\n",
      "  |                                                                            \n",
      "  |===========================================                           |  62%\n",
      "  |                                                                            \n",
      "  |=============================================                         |  65%\n",
      "  |                                                                            \n",
      "  |===============================================                       |  68%\n",
      "  |                                                                            \n",
      "  |=================================================                     |  71%\n",
      "  |                                                                            \n",
      "  |===================================================                   |  74%\n",
      "  |                                                                            \n",
      "  |======================================================                |  76%\n",
      "  |                                                                            \n",
      "  |========================================================              |  79%\n",
      "  |                                                                            \n",
      "  |==========================================================            |  82%\n",
      "  |                                                                            \n",
      "  |============================================================          |  85%\n",
      "  |                                                                            \n",
      "  |==============================================================        |  88%\n",
      "  |                                                                            \n",
      "  |================================================================      |  91%\n",
      "  |                                                                            \n",
      "  |==================================================================    |  94%\n",
      "  |                                                                            \n",
      "  |====================================================================  |  97%\n",
      "  |                                                                            \n",
      "  |======================================================================| 100%\n",
      "R[write to console]: \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Running PCA.\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: PC_ 1 \n",
      "Positive:  LAMA2, RYR2, TTN, PTPRG, MAML2, FBXL7, GRIK2, PRKG1, RORA, ZFPM2 \n",
      "\t   SLC8A1, ABCA9-AS1, CTNNA3, ABCA9, CBLB, ZEB1, BICC1, CACNA1C, AP002518.2, SOX5 \n",
      "\t   EBF1, LRRTM4, KAZN, ARHGAP10, FGF12, PDE3A, NAV2, AC007319.1, ACSM3, ARL15 \n",
      "Negative:  CALM2, RPL18A, CD59, RPL26, RPL39, OAZ1, RPS27, RPL21, RPS18, RPL9 \n",
      "\t   RPL7, RPL3, RPL34, TMSB4X, RPL32, RPS14, FTL, COX6A1, RPL12, RPS19 \n",
      "\t   HSPE1, NEDD8, LGALS1, VAMP2, B2M, FKBP1A, RPS15A, RPS15, ARL2, RPL23A \n",
      "PC_ 2 \n",
      "Positive:  PID1, ACSM3, VIT, DCLK1, EBF2, EGFR, ACSM1, FOXP2, MGST1, PLA2G5 \n",
      "\t   COL4A4, ADH1B, CARMIL1, EBF1, PDGFD, C7, LHFPL6, COL6A3, KAZN, PRKG1 \n",
      "\t   BNC2, AC007319.1, MSC-AS1, ABCA10, SOBP, FMO2, COL4A3, MECOM, UNC5C, ARHGAP10 \n",
      "Negative:  NPPA, CD9, NPPB, CRYAB, SPP1, PKN2-AS1, PMP22, GPX3, CLU, IL1RAPL2 \n",
      "\t   NFATC2, ALDH1A1, TMEM176A, HES1, S100A4, PLP1, CST3, MPZ, S100A6, TNNI3 \n",
      "\t   HS3ST3A1, UBASH3B, S100A1, CD14, LRRTM4, TNNT2, ALK, MYOT, RHOB, S100B \n",
      "PC_ 3 \n",
      "Positive:  CLDN5, TMEM88, BST2, CXCL2, ICAM2, SOX17, CAVIN2, GIMAP7, TM4SF1, ESAM \n",
      "\t   AQP1, CAV1, IL32, SRGN, FABP4, HLA-DRB5, FLT1, DLL4, GIMAP4, VWF \n",
      "\t   ECSCR, FCN3, PECAM1, GYPC, HLA-DPA1, TINAGL1, CD74, RNASE1, IFI27, CLEC14A \n",
      "Negative:  GPX3, CLU, PLP1, PID1, VWA1, FXYD1, TMEM176A, S100B, VIT, RARRES2 \n",
      "\t   C1R, CRYAB, S100A4, CST3, VIM, S100A6, PRNP, PDLIM2, FOXP2, LGALS1 \n",
      "\t   DCLK1, ACSM3, COL6A3, PMP22, ALDH1A1, ACSM1, ARHGEF10, SERPING1, KAZN, SPP1 \n",
      "PC_ 4 \n",
      "Positive:  PID1, FMO2, ACSM1, VIT, DCN, COL6A3, BICC1, TM4SF1, MECOM, BST2 \n",
      "\t   CLDN5, DCLK1, EGFR, TMEM88, ADH1B, COL4A4, C7, ICAM2, GIMAP7, FOXP2 \n",
      "\t   SPARCL1, SPRY1, COL4A3, CD34, CXCL2, CALCRL, ACSM3, ELN, CAVIN2, EBF1 \n",
      "Negative:  LGR5, GALNT17, LINC00639, AL450332.1, PPP2R2B, NKAIN2, KCNJ3, THSD7A, ARHGEF4, LPL \n",
      "\t   KCNE4, AC125613.1, ADAMTSL1, AC018742.1, RIMS1, LSAMP, GRIN2B, AC073050.1, COL11A1, MAPK4 \n",
      "\t   TRPM3, ADGRV1, RASEF, GPM6A, NRG3, SHANK2, ARHGEF28, KIAA1549L, EDIL3, MDGA2 \n",
      "PC_ 5 \n",
      "Positive:  LSAMP, LGR5, ABCA10, NKAIN2, THSD7A, PPP2R2B, TRPM3, AC018742.1, LINC00639, KAZN \n",
      "\t   ABCA9, ARHGEF4, AL450332.1, GRIN2B, NRG3, GPM6A, ABCA9-AS1, GRIK2, BNC2, MGST1 \n",
      "\t   ADAMTSL1, KCNE4, COL11A1, ADGRV1, AC125613.1, SHANK2, SYT10, VIT, KIAA1549L, LRRTM4 \n",
      "Negative:  CACNA1C, CD36, CARMN, GUCY1A2, RYR2, FHL2, LDB3, SLC8A1, ABCC9, RCAN2 \n",
      "\t   CACNB2, PTH1R, ADCY5, GUCY1A1, INPP4B, AC008250.1, AC131025.2, CCDC141, PDE3A, SLC8A1-AS1 \n",
      "\t   PLCL1, TTN, NOTCH3, RBM20, MLIP, EDNRA, MAST4, MYO1B, MYBPC3, MYOM2 \n",
      "\n",
      "R[write to console]: Computing nearest neighbor graph\n",
      "\n",
      "R[write to console]: Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 3436\n",
      "Number of edges: 109553\n",
      "\n",
      "Running Louvain algorithm...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Maximum modularity in 10 random starts: 0.8468\n",
      "Number of communities: 12\n",
      "Elapsed time: 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Warning:\n",
      "R[write to console]:  The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
      "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
      "This message will be shown once per session\n",
      "\n",
      "R[write to console]: 22:04:34 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "R[write to console]: Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "R[write to console]: Also defined by ‘spam’\n",
      "\n",
      "R[write to console]: 22:04:34 Read 3436 rows and found 10 numeric columns\n",
      "\n",
      "R[write to console]: 22:04:34 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "R[write to console]: Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "R[write to console]: Also defined by ‘spam’\n",
      "\n",
      "R[write to console]: 22:04:34 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n",
      "R[write to console]: 22:04:34 Writing NN index file to temp file /tmp/RtmpGF6ZfT/file43873e51a44\n",
      "\n",
      "R[write to console]: 22:04:34 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "R[write to console]: 22:04:35 Annoy recall = 100%\n",
      "\n",
      "R[write to console]: 22:04:36 Commencing smooth kNN distance calibration using 1 thread\n",
      "R[write to console]:  with target n_neighbors = 30\n",
      "\n",
      "R[write to console]: 22:04:37 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "R[write to console]: 22:04:37 Commencing optimization for 500 epochs, with 144658 positive edges\n",
      "\n",
      "R[write to console]: Using method '\n",
      "R[write to console]: umap\n",
      "R[write to console]: '\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n",
      "R[write to console]: 22:04:41 Optimization finished\n",
      "\n",
      "R[write to console]: Creating h5Seurat file for version 3.1.5.9900\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Seurat object is saved successfully.\"\n",
      "[1] \"Converting Seurat object to AnnData object.\"\n"
     ]
    }
   ],
   "source": [
    "ro.r['source']('seurat_qc.R')\n",
    "RunSeuratQC_r = ro.globalenv['RunSeuratQC']\n",
    "path=\"hca_heart_neuronal_raw.h5ad\"\n",
    "adata_path = 'hca_heart_neuronal_raw.h5ad'\n",
    "results = list(RunSeuratQC_r(path, 'hca_heart_neuronal_raw.h5seurat', adata_path=adata_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bf9cb63e-9bb4-4ad8-a1fd-0ebe4036a296",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[array(['RNA'], dtype='<U3'), array(['RNA'], dtype='<U3'), array(['hca_heart_neuronal_raw.h5seurat'], dtype='<U31'), array(['hca_heart_neuronal_raw.h5ad'], dtype='<U27'), <rpy2.robjects.vectors.BoolVector object at 0x7f69ec577600> [RTYPES.LGLSXP]\n",
      "R classes: ('logical',)\n",
      "[       0]]\n"
     ]
    }
   ],
   "source": [
    "print(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "11fd8c39-1a90-4936-8a86-4ebaf409ba09",
   "metadata": {},
   "outputs": [],
   "source": [
    "from rpy2.robjects import r, StrVector, NULL\n",
    "from typing import Any, List, Optional\n",
    "\n",
    "def convert_from_r(item: Any, date_cols: Optional[List[str]] = None, name: str = '', reserve_plots: bool = True) -> Any:\n",
    "    \"\"\"convert rpy object into python native object\n",
    "    Args:\n",
    "        item (Any): rpy2 object to convert\n",
    "        date_cols (list[str], optional): define the date colums in R dataframe , in order to conevrt them into pandas datetime. Defaults to None.\n",
    "        name (str, optional): name of the object to convert, (not required for external use). Defaults to ''.\n",
    "        reserve_plots (bool, optional): if True prserve rpy2 ListVector as rpy2 ListVector if name conatains plot,\n",
    "                                        in order to be able to ouput ggplot plots. Defaults to True.\n",
    "    Returns:\n",
    "        Any: the converted item\n",
    "    \"\"\"\n",
    "    result = item\n",
    "    remove_list: bool = True\n",
    "    if item == ro.vectors.NULL:\n",
    "        return None\n",
    "    elif 'plot' in name and isinstance(item, ro.vectors.ListVector) and reserve_plots:\n",
    "        return item\n",
    "    elif isinstance(item, (ro.environments.Environment,\n",
    "                           ro.Formula)):\n",
    "        return None\n",
    "    elif isinstance(item, ro.vectors.DataFrame):\n",
    "        with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "            result = ro.conversion.rpy2py(item)\n",
    "        result = convert_df_dates_from_r(result, date_cols)\n",
    "        remove_list = False\n",
    "    elif isinstance(item, (ro.vectors.StrVector,\n",
    "                           ro.vectors.FloatVector,\n",
    "                           ro.vectors.BoolVector,\n",
    "                           ro.vectors.IntVector)):\n",
    "        result = tuple(item)\n",
    "    elif isinstance(item, ro.vectors.ListVector):\n",
    "        if item.names == ro.vectors.NULL:\n",
    "            return None\n",
    "        result = {}\n",
    "        remove_list = False\n",
    "        if len(item) > 0:\n",
    "            result = dict(zip(item.names, list(item)))\n",
    "            for k, v in result.items():\n",
    "                result[k] = convert_from_r(v, date_cols, name=k)\n",
    "    if '__len__' in result.__dir__() and len(result) == 1 and remove_list:\n",
    "        result = result[0]\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5622a99f-cbcf-46d7-853f-b042177d6adc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Type of ddl_assay_names: <class 'bool'>\n",
      "Value of ddl_assay_names: False\n"
     ]
    }
   ],
   "source": [
    "if results[0] != ro.rinterface.NULL:\n",
    "    default_assay = list(results[0])[0]\n",
    "    assay_names = list(results[1])\n",
    "    output = list(results[2])[0]\n",
    "    adata_path = list(results[3])[0]\n",
    "    ddl_assay_names = convert_from_r(list(results[4])[0])\n",
    "    print(\"Type of ddl_assay_names:\", type(ddl_assay_names))\n",
    "    print(\"Value of ddl_assay_names:\", ddl_assay_names)\n",
    "else:\n",
    "    raise RuntimeError(f\"Seurat QC failed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "67133f9d-92ff-42e8-9083-48989e8bcd3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'hca_heart_neuronal_raw.h5seurat'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e9b56550-ac39-4932-9202-8c23baf6cf4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'hca_heart_neuronal_raw.h5ad'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "359ab467-39c8-4a26-9539-4f69b36925d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3436 × 33538\n",
       "    obs: 'orig.ident', 'n_counts', 'n_genes', 'nCounts_RNA', 'nFeatures_RNA', 'NRP', 'age_group', 'cell_source', 'cell_type', 'donor', 'gender', 'percent_mito', 'percent_ribo', 'region', 'sample', 'scrublet_score', 'source', 'type', 'version', 'cell_states', 'Used', 'pct_counts_mt', 'pct_counts_rb', 'pct_counts_hb', 'percent.plat', 'RNA_snn_res.0.5', 'seurat_clusters', 'doublet_score', 'doublet_class'\n",
       "    var: 'vst.mean', 'vst.variance', 'vst.variance.expected', 'vst.variance.standardized', 'highly_variable'\n",
       "    obsm: 'X_pca', 'X_tsne', 'X_umap'\n",
       "    layers: 'scale.data'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(adata_path)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "13c93014-a438-4a6e-a716-a381228a51d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[0., 0., 0., ..., 0., 0., 0.],\n",
       "        [0., 0., 0., ..., 0., 0., 0.],\n",
       "        [0., 0., 0., ..., 0., 0., 0.],\n",
       "        ...,\n",
       "        [0., 0., 0., ..., 0., 0., 0.],\n",
       "        [0., 0., 0., ..., 0., 0., 0.],\n",
       "        [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.todense()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ed33b2b0-25b7-4aa7-bf16-b6431f775142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 0.        ,  0.        , -0.0281569 , ..., -0.06304653,\n",
       "         0.        ,  0.        ],\n",
       "       [ 0.        ,  0.        , -0.0281569 , ..., -0.06304653,\n",
       "         0.        ,  0.        ],\n",
       "       [ 0.        ,  0.        , -0.0281569 , ..., -0.06304653,\n",
       "         0.        ,  0.        ],\n",
       "       ...,\n",
       "       [ 0.        ,  0.        , -0.0281569 , ..., -0.06304653,\n",
       "         0.        ,  0.        ],\n",
       "       [ 0.        ,  0.        , -0.0281569 , ..., -0.06304653,\n",
       "         0.        ,  0.        ],\n",
       "       [ 0.        ,  0.        , -0.0281569 , ..., -0.06304653,\n",
       "         0.        ,  0.        ]])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.layers['scale.data']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a5593dbd-e0ad-4078-8d1e-a824ea555d81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding 3D UMAP to AnnData object.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.8/dist-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.\n",
      "  warn(f\"n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.\")\n"
     ]
    }
   ],
   "source": [
    "from umap import UMAP\n",
    "\n",
    "if not ddl_assay_names:\n",
    "    if os.path.exists(adata_path):\n",
    "        print(\"Adding 3D UMAP to AnnData object.\")\n",
    "        sc.pp.neighbors(adata, n_neighbors=10, n_pcs=3, random_state=0)\n",
    "        umap_3d = UMAP(n_components=3, init='random', random_state=0)\n",
    "        adata.obsm[\"X_umap_3D\"] = umap_3d.fit_transform(adata.obsm['X_pca'])\n",
    "        # adata.obsm[\"X_umap_3D\"] = adata_3D.obsm[\"X_umap\"]\n",
    "        adata.write_h5ad(adata_path, compression='gzip')\n",
    "        adata_3D = None\n",
    "    else:\n",
    "        raise ValueError(\"AnnData file does not exist.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "14bc9d49-338b-40b5-bbfa-b63e3f8cf16c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3436 × 33538\n",
       "    obs: 'orig.ident', 'n_counts', 'n_genes', 'NRP', 'age_group', 'cell_source', 'cell_type', 'donor', 'gender', 'nCounts_RNA', 'nFeatures_RNA', 'percent_mito', 'percent_ribo', 'region', 'sample', 'scrublet_score', 'source', 'type', 'version', 'cell_states', 'Used', 'pct_counts_mt', 'pct_counts_rb', 'pct_counts_hb', 'percent.plat', 'RNA_snn_res.0.5', 'seurat_clusters', 'doublet_score', 'doublet_class'\n",
       "    var: 'vst.mean', 'vst.variance', 'vst.variance.expected', 'vst.variance.standardized', 'highly_variable'\n",
       "    uns: 'neighbors'\n",
       "    obsm: 'X_pca', 'X_tsne', 'X_umap', 'X_umap_3D'\n",
       "    layers: 'scale.data'\n",
       "    obsp: 'distances', 'connectivities'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "84c48aaa-db2a-4a0b-8e81-b62491009693",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 4.73994233, 10.88068801],\n",
       "       [ 0.05498769, -6.39408511],\n",
       "       [-0.81740091, -5.91939848],\n",
       "       ...,\n",
       "       [-8.59770701,  5.80163509],\n",
       "       [-9.05568335,  5.64351732],\n",
       "       [-8.73765872,  5.65475685]])"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obsm['X_umap']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "2ed29d4d-6546-47c4-a868-d595952e4d1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 1.2389023 ,  0.17247425,  3.5108423 ],\n",
       "       [-1.5085722 ,  1.0367727 ,  4.0758543 ],\n",
       "       [-1.4218253 ,  0.23663785,  3.2739491 ],\n",
       "       ...,\n",
       "       [14.681225  , -0.38077277,  8.828661  ],\n",
       "       [14.738458  , -0.43038186,  8.868302  ],\n",
       "       [14.769193  , -0.45038512,  8.847239  ]], dtype=float32)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obsm['X_umap_3D']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05c5ad00-7067-4da1-b51e-799939d78e1e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
